src/FwdState.cc

```diff
@@ -421,9 +421,11 @@ FwdState::Start(const Comm::ConnectionPointer &clientConn, StoreEntry *entry, Ht
         CacheManager::GetInstance()->start(clientConn, request, entry, al);
         return;
 
+#if USE_URN_THTTP
     case AnyP::PROTO_URN:
         urnStart(request, entry, al);
         return;
+#endif
```

Comment:
> > We have been down this road before. There is a process for deprecation of features - it begins with letting vendors choose whether they are going to drop/keep the code (this PR), and waiting for feedback (1-2 release cycles).
> 
> That old process has not worked well for years. We are wasting significant amounts of precious resources on features that virtually nobody needs instead of working on things that virtually everybody uses (or many folks need). It makes no sense to keep doing what harms us!

This process was last used with --disable-dns-internal was added in v3.2 to deprecate the helper ... nobody touched the deprecated code for years ... it was erased in v3.5 *instead* of being touched to support the v4 protocol changes.

The time before that was --enable-ntlm-fail-open being dropped. Again no problems with that. It was added some time ago and erased *instead* of fixing build issues when the helper code was refactored.

IME the main issue we have with dropping old code is that is often objections to *how* the removal is performed and we get into long time consuming threads like this one. Perhapse try letting others remove code without imposing your preference of methodology?

Reply:
> Alex: That old process has not worked well for years.

> Amos: This process was last used with --disable-dns-internal
> Amos: The time before that was --enable-ntlm-fail-open being dropped

I am talking about the process of disabling and/or deprecating certain features (in master) instead of simply removing them (from master). There are lots of features in Squid master today that we are wasting precious resources on without sufficient payoff. Adding URNs to that list is wrong IMO. The two examples above are not representative of the widespread problem I am talking about, even if the process worked perfectly in every way for those two examples.

> IME the main issue we have with dropping old code is that is often objections to _how_ the removal is performed and we get into long time consuming threads like this one. Perhapse try letting others remove code without imposing your preference of methodology?

Besides suggesting to completely remove URNs (something that is easy to do and something that would reduce long-term waste for all of us), I have requested an objective PR improvement and even volunteered to supply it. Carefully evaluating suggestions on their merits would probably prevent most of Squid PR waste and pain. For example, it would shorten this change request thread to one comment (if URNs are removed from master) or two comments (if URNs are disabled-by-default without adding `#ifdefs` to FwdState and friends).

Reply:
> > Alex: That old process has not worked well for years.
> 
> > Amos: This process was last used with --disable-dns-internal
> > Amos: The time before that was --enable-ntlm-fail-open being dropped
> 
> I am talking about the process of disabling and/or deprecating certain features (in master) instead of simply removing them (from master).

So am I. The process this PR follows is standard practice and works well as I have referenced. The avoidance of immediate removal is an explicit design point with good reason to be adhered to.

There is a high potential some unknown users to still be relying on the feature *and* not tracking what we are doing - they need a quick way to request a halt to the deprecation when they reach the version where a feature is disabled (removed from vendors builds).

> There are lots of features in Squid master today that we are wasting precious resources on without sufficient payoff. 

The code this PR is about, is not one of those. Except that you are making all this unnecessary discussion.

> Adding URNs to that list is wrong IMO. 

Then please remove your merge objection and we can let this URN stuff die away peacefully as we get on to the other features and code updates that need doing.

>The two examples above are not representative of the widespread problem I am talking about, even if the process worked perfectly in every way for those two examples.

Those are the only feature removals which correlate to this process since v3. All other feature removals have been upgrade migrations where there was another feature for users to switch to - not complete removals.

> 
> > IME the main issue we have with dropping old code is that is often objections to _how_ the removal is performed and we get into long time consuming threads like this one. Perhapse try letting others remove code without imposing your preference of methodology?
> 
> Besides suggesting to completely remove URNs (something that is easy to do and something that would reduce long-term waste for all of us), I have requested an objective PR improvement and even volunteered to supply it.

As I replied to that, your suggested "improvement" is far from an improvement. It actively adds a huge amount of ongoing maintenance requirements in place of this simple removal process. You are effectively requiring use to take on support for this IETF deprecated feature and maintain it long-term without any use-case justification.

> Carefully evaluating suggestions on their merits would probably prevent most of Squid PR waste and pain.
> For example, it would shorten this change request thread to one comment (if URNs are removed from master) or two comments (if URNs are disabled-by-default without adding `#ifdefs` to FwdState and friends).

You imply that careful evaluation was not done. see https://github.com/squid-cache/squid/pull/915#discussion_r735199369 for the relative merits. 

The design of the process followed ensures a) removal of a feature, and b) avoid irreversibly breaking unknown users installations, and c) with minimal code and/or code maintenance requirements added for us. Your two suggestions violate (b) and (c) respectively.

In the end this request would still be a request to change an *industry standard process* by adding unnecessary extra harm/pain to either project dev or the project users. So, my answer to that is still "won't change".

Reply:
> The process this PR follows is standard practice and works well as I have referenced. The avoidance of immediate removal is an explicit design point with good reason to be adhered to.

The process clearly does not work well: There is a lot of stale stuff that is not removed and wastes our time. Besides, we cannot afford to support the good intentions behind that process design IMO. However, we have said all that already. I am not sure why you are revisiting this without bringing any new points (that I can see).


> There is a high potential some unknown users to still be relying on the feature _and_ not tracking what we are doing

Based on poor code quality and the lack of references on public Squid channels, I doubt the "high potential" claim is true for URNs, but it does not really matter: We have to cut maintenance costs to move forward, and URNs are a great candidate for doing so. They waste significant resources on ongoing basis, and their poor implementation quality creates CVEs (wasting even more resources in high-priority mode).


> they need a quick way to request a halt to the deprecation when they reach the version where a feature is disabled (removed from vendors builds).

Again, this is one of the (hypothetical) needs we should not satisfy. I am not disputing (and have never disputed) that (hypothetical) need. There is no need to spend time convincing me that it may exist.


> > There are lots of features in Squid master today that we are wasting precious resources on without sufficient payoff.
> 
> The code this PR is about, is not one of those. Except that you are making all this unnecessary discussion.

My change requests were specific to URNs. I mentioned other features/subjects only in response to arguments that used/introduced other features/subjects. I did not request any discussions -- my change requests could be addressed without any discussion at all.


> > Adding URNs to that list is wrong IMO.

> Then please remove your merge objection and we can let this URN stuff die away peacefully as we get on to the other features and code updates that need doing.

I see no logical connection between my statement that URNs should be removed rather than disabled-by-default (due to long-term maintenance costs) and your statement that disabling URNs by default (using macros) will somehow allow them to die peacefully, but perhaps I just cannot guess what "die peacefully" means.
 
> your suggested "improvement" is far from an improvement. It actively adds a huge amount of ongoing maintenance requirements in place of this simple removal process.

It does not.

> You are effectively requiring use to take on support for this IETF deprecated feature and maintain it long-term without any use-case justification.

I am not. It sounds like you are talking about something I have not suggested, but I do not know what that is.


> > Carefully evaluating suggestions on their merits would probably prevent most of Squid PR waste and pain.
> > For example, it would shorten this change request thread to one comment (if URNs are removed from master) or two comments (if URNs are disabled-by-default without adding `#ifdefs` to FwdState and friends).
 
> You imply that careful evaluation was not done. see [#915 (comment)](https://github.com/squid-cache/squid/pull/915#discussion_r735199369) for the relative merits.

AFAICT, that comment evaluated something I have not suggested.

Reply:
> Alex: It is possible to address this change request without requiring more UI changes than this PR already requires and without changing config-related code.

> Amos: Please do enlighten me on how one is expected to implement run-time checks of admin desires without a configuration option in cf.data.pre, code to check that configuration option, nor any alterations to cache_cf.cc?

Done in PR branch commit ef49a55.

That commit has a nice side effect: It fixes another PR problem that earlier reviews of `#if`ed code have missed.

Reply:
Your ef49a559cff0756c46d1cbb6b4563695c8a91265 changes converted relayed "urn://" messages into a fatal shutdown of Squid when they were not Trivial-HTTP resolvable. This is not acceptable.

Closing again as complete removal of Trivial-HTTP is now happening.

Reply:
> Your https://github.com/squid-cache/squid/commit/ef49a559cff0756c46d1cbb6b4563695c8a91265 changes converted relayed "urn://" messages into a fatal shutdown of Squid when they were not Trivial-HTTP resolvable.

The above assertion is false. Compared to official code, commit ef49a559cff0756c46d1cbb6b4563695c8a91265 did not add any _new_ code paths that would end in fatal shutdowns. The existing/preserved code path that ends at fatal_dump() in FwdState::dispatch() is _unreachable_ code. "not Trivial-HTTP resolvable" requests result in error responses generated/stored by urn.cc code without using FwdState::dispatch().